src/random.c

```diff
@@ -0,0 +1,108 @@
+/* Copyright libuv contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "uv-common.h"
+
+#ifdef _WIN32
+extern int uv__random_rtlgenrandom(void* buf, size_t buflen);
+#else
+#include "unix/internal.h"
+#endif
+
+
+static int uv__random(void* buf, size_t buflen) {
+  int rc;
+
+#if defined(__APPLE__) || defined(__OpenBSD__)
+  rc = uv__random_getentropy(buf, buflen);
+  if (rc == UV_ENOSYS)
+    rc = uv__random_devurandom(buf, buflen);
+#elif defined(__FreeBSD__) || defined(__linux__)
+  rc = uv__random_getrandom(buf, buflen);
+  if (rc == UV_ENOSYS)
+    rc = uv__random_devurandom(buf, buflen);
+# if defined(__linux__)
+  if (rc == UV_ENOENT)
+    rc = uv__random_sysctl(buf, buflen);
+  /* TODO(bnoordhuis) Try parsing /proc/sys/kernel/random/uuid? */
```

Comment:
Is the `/proc/sys/kernel/random/uuid` interface the same used by the `sysctl()` call? If so, wouldn't it be worth just using the former as `sysctl()` calls have been deprecated for a while?

Reply:
I added this code path for systems that:

1. don't have getrandom(), and
2. doesn't have (a readable) /dev/urandom, and
3. where procfs isn't mounted.

I figured a deprecated system call is an acceptable trade-off because this code path will only be used on non-standard systems with old kernels.

(Looking at this code again I should probably check for more than just ENOENT. EACCES, ELOOP, EMFILE, etc., should also trigger the fallback.)